### PR TITLE
SOLR-17153 Resolve IndexOutOfBoundsException

### PR DESCRIPTION
**Pull request by aparnasuresh85**
https://issues.apache.org/jira/browse/SOLR-17153

# Description

Resolve an IndexOutOfBoundsException in HttpSolrCall.resolveDocCollection(), at line: 
String collectionName = collectionsList.get(0);

# Solution

Check if collectionsList is empty, in that case return early. 

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [X] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [X] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [X] I have developed this patch against the `main` branch.
- [X] I have run `./gradlew check`.
- [] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java [hunk 1]
```diff
@@ -354,13 +352,15 @@ protected void init() throws Exception {
    * recursively try this all one more time while ensuring the alias and collection info is sync'ed
    * from ZK.
    */
-  protected DocCollection resolveDocCollection(List<String> collectionsList) {
+  protected DocCollection resolveDocCollection(String collectionName) {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java:355]: Its javadocs appear outdated now.

I like the signature change.

File: solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java [hunk 2]
```diff
@@ -354,13 +352,15 @@ protected void init() throws Exception {
...
     if (!cores.isZooKeeperAware()) {
       throw new SolrException(
           SolrException.ErrorCode.BAD_REQUEST, "Solr not running in cloud mode ");
     }
+    if (collectionName == null || collectionName.trim().isEmpty()) {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java:360]:
```suggestion
    if (collectionName == null || collectionName.isBlank()) {
```

File: solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java [hunk 3]
```diff
@@ -1066,13 +1066,14 @@ private Map<String, Integer> checkStateVersionsAreValid(String stateVer) {
 
   protected SolrCore getCoreByCollection(String collectionName, boolean isPreferLeader) {
     ZkStateReader zkStateReader = cores.getZkController().getZkStateReader();
-
     ClusterState clusterState = zkStateReader.getClusterState();
-    DocCollection collection = clusterState.getCollectionOrNull(collectionName, true);
+    DocCollection collection = resolveDocCollection(collectionName);
+    // the usage of getCoreByCollection assumes that if null is returned, collection is found, but
+    // replicas might not
+    // have been created. Hence returning null here would be misleading...
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java:1073]: Seems like you could write javadocs for getCoreByCollection additionally or instead

**Review by dsmiley — COMMENTED** [review]
If Eric found a bug then it seems we need a test for this bug

**Comment by aparnasuresh85** [comment]
> If Eric found a bug then it seems we need a test for this bug

I had the same thought. But I now see the line of code that caused it is no longer being used. The call _resolveDocCollection()_ is now being invoked within _getCoreByCollection()_ and its signature has also changed.

What would be the test case of the new test? We would need one that doesnt create a collection or noneof the usual ones. Eric tried to start solr and launch the admin page.....

**Comment by epugh** [comment]
I think a test of a Solr with no collections created should demonstrate the bug.  Please let me know if I can help.  I'd love to see this fixed so I can then merge some other code into `main`.

**Comment by aparnasuresh85** [comment]
> I think a test of a Solr with no collections created should demonstrate the bug. Please let me know if I can help. I'd love to see this fixed so I can then merge some other code into `main`.

Hi Eric,

I started working on a Java test which didnt really work as expected to address this use case (launch the solr admin console ). I am now looking to add a test case in a ".bats" script under solr/packaging. Let me know if this sounds appropriate.

File: solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java (hunk 3 above)
Comment (aparnasuresh85) [reply solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java:1079]: Done

File: solr/packaging/test/test_launch_solr_adminconsole.bats [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+@test "assert able to launch solr admin console" {
+  run solr start -c
+
+  # Replace 'YOUR_ADMIN_CONSOLE_URL' with the actual admin console URL
+  run curl -s -o /dev/null -w "%{http_code}" http://localhost:${SOLR_PORT}/solr/
+
+  # Check if the HTTP response code is 200 (OK)
+  assert_output --partial "200"
```
Comment (aparnasuresh85) [inline solr/packaging/test/test_launch_solr_adminconsole.bats:38]: The test failed with the bug and passed without it.

File: solr/packaging/test/test_launch_solr_adminconsole.bats [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+teardown() {
+  # save a snapshot of SOLR_HOME for failed tests
+  save_home_on_failure
+
+  solr stop -all >/dev/null 2>&1
+}
+
+@test "assert able to launch solr admin console" {
```
Comment (aparnasuresh85) [inline solr/packaging/test/test_launch_solr_adminconsole.bats:31]: Please suggest an alternative name for the .bats script if you feel it is too specific.

**Comment by aparnasuresh85** [comment]
> I think a test of a Solr with no collections created should demonstrate the bug. Please let me know if I can help. I'd love to see this fixed so I can then merge some other code into `main`.

Hi @epugh , @dsmiley ,

I have updated the PR with a .bats test. Please let me know your feedback.

File: solr/packaging/test/test_launch_solr_adminconsole.bats (hunk 4 above)
Comment (dsmiley) [reply solr/packaging/test/test_launch_solr_adminconsole.bats:37]: Is the `--partial` part accurate -- won't `200` be the entirety of the output?

File: solr/packaging/test/test_launch_solr_adminconsole.bats
[the remark this answers is not in the document]
Comment (aparnasuresh85) [reply solr/packaging/test/test_launch_solr_adminconsole.bats:1]: Done

File: solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java (hunk 1 above)
Comment (aparnasuresh85) [reply solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java:356]: Updated.

File: solr/packaging/test/test_launch_solr_adminconsole.bats (hunk 4 above)
Comment (aparnasuresh85) [reply solr/packaging/test/test_launch_solr_adminconsole.bats]: I modified the test to instead match exact output.

**Comment by dsmiley** [comment]
Looks good; I'd like to merge.  If you disagree @epugh , please pipe in today.

Minor nitpick:  there's a comment that doesn't flow (newline) nicely but whatever.

**Comment by epugh** [comment]
I just manually gave this PR a run through, and it looks great.   Please merge!

**Pull request merged**